#### Welcome to the third question of assignment.

# Question:3 : Explore and Cluster Toronto neighborhood

#### Here we continue from the merged dataframes for Toronto postal codes. T

### Converting html table to panda dataframe

#### The html table is first extracted from the Wikipedia page. The html table is stored in a variable html2. Using BeatifulSoup I am trying to convert HTML table into panda dataframe.


In [2]:
# Import required libraries
import pandas as pd
from bs4 import BeautifulSoup
print("Done")

Done


In [3]:

html2 = """
<table class="wikitable">
<tr>
<th>Postal code
</th>
<th>Borough
</th>
<th>Neighborhood
</th></tr>
<tr>
<td>M1A
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M2A
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3A
</td>
<td>North York
</td>
<td>Parkwoods
</td></tr>
<tr>
<td>M4A
</td>
<td>North York
</td>
<td>Victoria Village
</td></tr>
<tr>
<td>M5A
</td>
<td>Downtown Toronto
</td>
<td>Regent Park / Harbourfront
</td></tr>
<tr>
<td>M6A
</td>
<td>North York
</td>
<td>Lawrence Manor / Lawrence Heights
</td></tr>
<tr>
<td>M7A
</td>
<td>Downtown Toronto
</td>
<td>Queen's Park / Ontario Provincial Government
</td></tr>
<tr>
<td>M8A
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9A
</td>
<td>Etobicoke
</td>
<td>Islington Avenue
</td></tr>
<tr>
<td>M1B
</td>
<td>Scarborough
</td>
<td>Malvern / Rouge
</td></tr>
<tr>
<td>M2B
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3B
</td>
<td>North York
</td>
<td>Don Mills
</td></tr>
<tr>
<td>M4B
</td>
<td>East York
</td>
<td>Parkview Hill / Woodbine Gardens
</td></tr>
<tr>
<td>M5B
</td>
<td>Downtown Toronto
</td>
<td>Garden District, Ryerson
</td></tr>
<tr>
<td>M6B
</td>
<td>North York
</td>
<td>Glencairn
</td></tr>
<tr>
<td>M7B
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8B
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9B
</td>
<td>Etobicoke
</td>
<td>West Deane Park / Princess Gardens / Martin Grove / Islington / Cloverdale
</td></tr>
<tr>
<td>M1C
</td>
<td>Scarborough
</td>
<td>Rouge Hill / Port Union / Highland Creek
</td></tr>
<tr>
<td>M2C
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3C
</td>
<td>North York
</td>
<td>Don Mills
</td></tr>
<tr>
<td>M4C
</td>
<td>East York
</td>
<td>Woodbine Heights
</td></tr>
<tr>
<td>M5C
</td>
<td>Downtown Toronto
</td>
<td>St. James Town
</td></tr>
<tr>
<td>M6C
</td>
<td>York
</td>
<td>Humewood-Cedarvale
</td></tr>
<tr>
<td>M7C
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8C
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9C
</td>
<td>Etobicoke
</td>
<td>Eringate / Bloordale Gardens / Old Burnhamthorpe / Markland Wood
</td></tr>
<tr>
<td>M1E
</td>
<td>Scarborough
</td>
<td>Guildwood / Morningside / West Hill
</td></tr>
<tr>
<td>M2E
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3E
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4E
</td>
<td>East Toronto
</td>
<td>The Beaches
</td></tr>
<tr>
<td>M5E
</td>
<td>Downtown Toronto
</td>
<td>Berczy Park
</td></tr>
<tr>
<td>M6E
</td>
<td>York
</td>
<td>Caledonia-Fairbanks
</td></tr>
<tr>
<td>M7E
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8E
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9E
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M1G
</td>
<td>Scarborough
</td>
<td>Woburn
</td></tr>
<tr>
<td>M2G
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3G
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4G
</td>
<td>East York
</td>
<td>Leaside
</td></tr>
<tr>
<td>M5G
</td>
<td>Downtown Toronto
</td>
<td>Central Bay Street
</td></tr>
<tr>
<td>M6G
</td>
<td>Downtown Toronto
</td>
<td>Christie
</td></tr>
<tr>
<td>M7G
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8G
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9G
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M1H
</td>
<td>Scarborough
</td>
<td>Cedarbrae
</td></tr>
<tr>
<td>M2H
</td>
<td>North York
</td>
<td>Hillcrest Village
</td></tr>
<tr>
<td>M3H
</td>
<td>North York
</td>
<td>Bathurst Manor / Wilson Heights / Downsview North
</td></tr>
<tr>
<td>M4H
</td>
<td>East York
</td>
<td>Thorncliffe Park
</td></tr>
<tr>
<td>M5H
</td>
<td>Downtown Toronto
</td>
<td>Richmond / Adelaide / King
</td></tr>
<tr>
<td>M6H
</td>
<td>West Toronto
</td>
<td>Dufferin / Dovercourt Village
</td></tr>
<tr>
<td>M7H
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8H
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9H
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M1J
</td>
<td>Scarborough
</td>
<td>Scarborough Village
</td></tr>
<tr>
<td>M2J
</td>
<td>North York
</td>
<td>Fairview / Henry Farm / Oriole
</td></tr>
<tr>
<td>M3J
</td>
<td>North York
</td>
<td>Northwood Park / York University
</td></tr>
<tr>
<td>M4J
</td>
<td>East York
</td>
<td>East Toronto
</td></tr>
<tr>
<td>M5J
</td>
<td>Downtown Toronto
</td>
<td>Harbourfront East / Union Station / Toronto Islands
</td></tr>
<tr>
<td>M6J
</td>
<td>West Toronto
</td>
<td>Little Portugal / Trinity
</td></tr>
<tr>
<td>M7J
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8J
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9J
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M1K
</td>
<td>Scarborough
</td>
<td>Kennedy Park / Ionview / East Birchmount Park
</td></tr>
<tr>
<td>M2K
</td>
<td>North York
</td>
<td>Bayview Village
</td></tr>
<tr>
<td>M3K
</td>
<td>North York
</td>
<td>Downsview
</td></tr>
<tr>
<td>M4K
</td>
<td>East Toronto
</td>
<td>The Danforth West / Riverdale
</td></tr>
<tr>
<td>M5K
</td>
<td>Downtown Toronto
</td>
<td>Toronto Dominion Centre / Design Exchange
</td></tr>
<tr>
<td>M6K
</td>
<td>West Toronto
</td>
<td>Brockton / Parkdale Village / Exhibition Place
</td></tr>
<tr>
<td>M7K
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8K
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9K
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M1L
</td>
<td>Scarborough
</td>
<td>Golden Mile / Clairlea / Oakridge
</td></tr>
<tr>
<td>M2L
</td>
<td>North York
</td>
<td>York Mills / Silver Hills
</td></tr>
<tr>
<td>M3L
</td>
<td>North York
</td>
<td>Downsview
</td></tr>
<tr>
<td>M4L
</td>
<td>East Toronto
</td>
<td>India Bazaar / The Beaches West
</td></tr>
<tr>
<td>M5L
</td>
<td>Downtown Toronto
</td>
<td>Commerce Court / Victoria Hotel
</td></tr>
<tr>
<td>M6L
</td>
<td>North York
</td>
<td>North Park / Maple Leaf Park / Upwood Park
</td></tr>
<tr>
<td>M7L
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8L
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9L
</td>
<td>North York
</td>
<td>Humber Summit
</td></tr>
<tr>
<td>M1M
</td>
<td>Scarborough
</td>
<td>Cliffside / Cliffcrest / Scarborough Village West
</td></tr>
<tr>
<td>M2M
</td>
<td>North York
</td>
<td>Willowdale / Newtonbrook
</td></tr>
<tr>
<td>M3M
</td>
<td>North York
</td>
<td>Downsview
</td></tr>
<tr>
<td>M4M
</td>
<td>East Toronto
</td>
<td>Studio District
</td></tr>
<tr>
<td>M5M
</td>
<td>North York
</td>
<td>Bedford Park / Lawrence Manor East
</td></tr>
<tr>
<td>M6M
</td>
<td>York
</td>
<td>Del Ray / Mount Dennis / Keelsdale and Silverthorn
</td></tr>
<tr>
<td>M7M
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8M
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9M
</td>
<td>North York
</td>
<td>Humberlea / Emery
</td></tr>
<tr>
<td>M1N
</td>
<td>Scarborough
</td>
<td>Birch Cliff / Cliffside West
</td></tr>
<tr>
<td>M2N
</td>
<td>North York
</td>
<td>Willowdale
</td></tr>
<tr>
<td>M3N
</td>
<td>North York
</td>
<td>Downsview
</td></tr>
<tr>
<td>M4N
</td>
<td>Central Toronto
</td>
<td>Lawrence Park
</td></tr>
<tr>
<td>M5N
</td>
<td>Central Toronto
</td>
<td>Roselawn
</td></tr>
<tr>
<td>M6N
</td>
<td>York
</td>
<td>Runnymede / The Junction North
</td></tr>
<tr>
<td>M7N
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8N
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9N
</td>
<td>York
</td>
<td>Weston
</td></tr>
<tr>
<td>M1P
</td>
<td>Scarborough
</td>
<td>Dorset Park / Wexford Heights / Scarborough Town Centre
</td></tr>
<tr>
<td>M2P
</td>
<td>North York
</td>
<td>York Mills West
</td></tr>
<tr>
<td>M3P
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4P
</td>
<td>Central Toronto
</td>
<td>Davisville North
</td></tr>
<tr>
<td>M5P
</td>
<td>Central Toronto
</td>
<td>Forest Hill North &amp; West
</td></tr>
<tr>
<td>M6P
</td>
<td>West Toronto
</td>
<td>High Park / The Junction South
</td></tr>
<tr>
<td>M7P
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8P
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9P
</td>
<td>Etobicoke
</td>
<td>Westmount
</td></tr>
<tr>
<td>M1R
</td>
<td>Scarborough
</td>
<td>Wexford / Maryvale
</td></tr>
<tr>
<td>M2R
</td>
<td>North York
</td>
<td>Willowdale
</td></tr>
<tr>
<td>M3R
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4R
</td>
<td>Central Toronto
</td>
<td>North Toronto West
</td></tr>
<tr>
<td>M5R
</td>
<td>Central Toronto
</td>
<td>The Annex / North Midtown / Yorkville
</td></tr>
<tr>
<td>M6R
</td>
<td>West Toronto
</td>
<td>Parkdale / Roncesvalles
</td></tr>
<tr>
<td>M7R
</td>
<td>Mississauga
</td>
<td>Canada Post Gateway Processing Centre
</td></tr>
<tr>
<td>M8R
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9R
</td>
<td>Etobicoke
</td>
<td>Kingsview Village / St. Phillips / Martin Grove Gardens / Richview Gardens
</td></tr>
<tr>
<td>M1S
</td>
<td>Scarborough
</td>
<td>Agincourt
</td></tr>
<tr>
<td>M2S
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3S
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4S
</td>
<td>Central Toronto
</td>
<td>Davisville
</td></tr>
<tr>
<td>M5S
</td>
<td>Downtown Toronto
</td>
<td>University of Toronto / Harbord
</td></tr>
<tr>
<td>M6S
</td>
<td>West Toronto
</td>
<td>Runnymede / Swansea
</td></tr>
<tr>
<td>M7S
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8S
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9S
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M1T
</td>
<td>Scarborough
</td>
<td>Clarks Corners / Tam O'Shanter / Sullivan
</td></tr>
<tr>
<td>M2T
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3T
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4T
</td>
<td>Central Toronto
</td>
<td>Moore Park / Summerhill East
</td></tr>
<tr>
<td>M5T
</td>
<td>Downtown Toronto
</td>
<td>Kensington Market / Chinatown / Grange Park
</td></tr>
<tr>
<td>M6T
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M7T
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8T
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9T
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M1V
</td>
<td>Scarborough
</td>
<td>Milliken / Agincourt North / Steeles East / L'Amoreaux East
</td></tr>
<tr>
<td>M2V
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3V
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4V
</td>
<td>Central Toronto
</td>
<td>Summerhill West / Rathnelly / South Hill / Forest Hill SE / Deer Park
</td></tr>
<tr>
<td>M5V
</td>
<td>Downtown Toronto
</td>
<td>CN Tower / King and Spadina / Railway Lands / Harbourfront West / Bathurst
<pre> Quay / South Niagara / Island airport
</pre>
</td></tr>
<tr>
<td>M6V
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M7V
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8V
</td>
<td>Etobicoke
</td>
<td>New Toronto / Mimico South / Humber Bay Shores
</td></tr>
<tr>
<td>M9V
</td>
<td>Etobicoke
</td>
<td>South Steeles / Silverstone / Humbergate / Jamestown / Mount Olive / Beaumond Heights / Thistletown / Albion Gardens
</td></tr>
<tr>
<td>M1W
</td>
<td>Scarborough
</td>
<td>Steeles West / L'Amoreaux West
</td></tr>
<tr>
<td>M2W
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3W
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4W
</td>
<td>Downtown Toronto
</td>
<td>Rosedale
</td></tr>
<tr>
<td>M5W
</td>
<td>Downtown Toronto
</td>
<td>Stn A PO Boxes
</td></tr>
<tr>
<td>M6W
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M7W
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8W
</td>
<td>Etobicoke
</td>
<td>Alderwood / Long Branch
</td></tr>
<tr>
<td>M9W
</td>
<td>Etobicoke
</td>
<td>Northwest
</td></tr>
<tr>
<td>M1X
</td>
<td>Scarborough
</td>
<td>Upper Rouge
</td></tr>
<tr>
<td>M2X
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3X
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4X
</td>
<td>Downtown Toronto
</td>
<td>St. James Town / Cabbagetown
</td></tr>
<tr>
<td>M5X
</td>
<td>Downtown Toronto
</td>
<td>First Canadian Place / Underground city
</td></tr>
<tr>
<td>M6X
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M7X
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8X
</td>
<td>Etobicoke
</td>
<td>The Kingsway / Montgomery Road  / Old Mill North
</td></tr>
<tr>
<td>M9X
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M1Y
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M2Y
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3Y
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4Y
</td>
<td>Downtown Toronto
</td>
<td>Church and Wellesley
</td></tr>
<tr>
<td>M5Y
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M6Y
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M7Y
</td>
<td>East Toronto
</td>
<td>Business reply mail Processing CentrE
</td></tr>
<tr>
<td>M8Y
</td>
<td>Etobicoke
</td>
<td>Old Mill South / King's Mill Park / Sunnylea / Humber Bay / Mimico NE / The Queensway East / Royal York South East / Kingsway Park South East
</td></tr>
<tr>
<td>M9Y
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M1Z
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M2Z
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3Z
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M4Z
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M5Z
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M6Z
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M7Z
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M8Z
</td>
<td>Etobicoke
</td>
<td>Mimico NW / The Queensway West / South of Bloor / Kingsway Park South West / Royal York South West
</td></tr>
<tr>
<td>M9Z
</td>
<td>Not assigned
</td>
<td>
</td></tr>
</table>"""





#### Use BeutifulSoup to convert html to dataframe.

In [4]:
soup = BeautifulSoup(html2, 'html.parser')
table = soup.find("table")

# The first tr contains the field names.
headings = [th.get_text().strip() for th in table.find("tr").find_all("th")]


datasets = []
for row in table.find_all("tr")[1:]:
    dataset = dict(zip(headings, (td.get_text() for td in row.find_all("td"))))
    datasets.append(dataset)


df=pd.DataFrame(datasets)
df.head(20)

Borough                                       Neighborhood  \
0       Not assigned\n                                                 \n   
1       Not assigned\n                                                 \n   
2         North York\n                                        Parkwoods\n   
3         North York\n                                 Victoria Village\n   
4   Downtown Toronto\n                       Regent Park / Harbourfront\n   
5         North York\n                Lawrence Manor / Lawrence Heights\n   
6   Downtown Toronto\n     Queen's Park / Ontario Provincial Government\n   
7       Not assigned\n                                                 \n   
8          Etobicoke\n                                 Islington Avenue\n   
9        Scarborough\n                                  Malvern / Rouge\n   
10      Not assigned\n                                                 \n   
11        North York\n                                        Don Mills\n   
12         East York\n                 Parkview Hill / Woodbine Gardens\n   
13  Downtown Toronto\n                         Garden District, Ryerson\n   
14        North York\n                                        Glencairn\n   
15      Not assigned\n                                                 \n   
16      Not assigned\n                                                 \n   
17         Etobicoke\n  West Deane Park / Princess Gardens / Martin Gr...   
18       Scarborough\n         Rouge Hill / Port Union / Highland Creek\n   
19      Not assigned\n                                                 \n   

   Postal code  
0        M1A\n  
1        M2A\n  
2        M3A\n  
3        M4A\n  
4        M5A\n  
5        M6A\n  
6        M7A\n  
7        M8A\n  
8        M9A\n  
9        M1B\n  
10       M2B\n  
11       M3B\n  
12       M4B\n  
13       M5B\n  
14       M6B\n  
15       M7B\n  
16       M8B\n  
17       M9B\n  
18       M1C\n  
19       M2C\n

### Remove last two unwanted characters

In [5]:
df['Borough'] = df['Borough'].map(lambda x: str(x)[:-1])
df['Postal code'] = df['Postal code'].map(lambda x: str(x)[:-1])
df['Neighborhood'] = df['Neighborhood'].map(lambda x: str(x)[:-1])
df.head(30)

Borough                                       Neighborhood  \
0       Not assigned                                                      
1       Not assigned                                                      
2         North York                                          Parkwoods   
3         North York                                   Victoria Village   
4   Downtown Toronto                         Regent Park / Harbourfront   
5         North York                  Lawrence Manor / Lawrence Heights   
6   Downtown Toronto       Queen's Park / Ontario Provincial Government   
7       Not assigned                                                      
8          Etobicoke                                   Islington Avenue   
9        Scarborough                                    Malvern / Rouge   
10      Not assigned                                                      
11        North York                                          Don Mills   
12         East York                   Parkview Hill / Woodbine Gardens   
13  Downtown Toronto                           Garden District, Ryerson   
14        North York                                          Glencairn   
15      Not assigned                                                      
16      Not assigned                                                      
17         Etobicoke  West Deane Park / Princess Gardens / Martin Gr...   
18       Scarborough           Rouge Hill / Port Union / Highland Creek   
19      Not assigned                                                      
20        North York                                          Don Mills   
21         East York                                   Woodbine Heights   
22  Downtown Toronto                                     St. James Town   
23              York                                 Humewood-Cedarvale   
24      Not assigned                                                      
25      Not assigned                                                      
26         Etobicoke  Eringate / Bloordale Gardens / Old Burnhamthor...   
27       Scarborough                Guildwood / Morningside / West Hill   
28      Not assigned                                                      
29      Not assigned                                                      

   Postal code  
0          M1A  
1          M2A  
2          M3A  
3          M4A  
4          M5A  
5          M6A  
6          M7A  
7          M8A  
8          M9A  
9          M1B  
10         M2B  
11         M3B  
12         M4B  
13         M5B  
14         M6B  
15         M7B  
16         M8B  
17         M9B  
18         M1C  
19         M2C  
20         M3C  
21         M4C  
22         M5C  
23         M6C  
24         M7C  
25         M8C  
26         M9C  
27         M1E  
28         M2E  
29         M3E

### Change column order

In [6]:
#change column order
df = df[["Postal code", "Borough", "Neighborhood"]] 
df.head()

Postal code           Borough                Neighborhood
0         M1A      Not assigned                            
1         M2A      Not assigned                            
2         M3A        North York                   Parkwoods
3         M4A        North York            Victoria Village
4         M5A  Downtown Toronto  Regent Park / Harbourfront

### Rename column

In [7]:
#rename column
df.rename(columns = {'Postal code':'PostalCode'}, inplace = True) 
df.head()


PostalCode           Borough                Neighborhood
0        M1A      Not assigned                            
1        M2A      Not assigned                            
2        M3A        North York                   Parkwoods
3        M4A        North York            Victoria Village
4        M5A  Downtown Toronto  Regent Park / Harbourfront

### Filter rows with Borough "Not assigned"

In [8]:
#filter rows with Borough not assigned
print(df.shape)
df = df[df.Borough != "Not assigned"]
print(df.shape)
df = df.reset_index(drop=True)
print(df.shape)
df.head()

(180, 3)
(103, 3)
(103, 3)


PostalCode           Borough                                  Neighborhood
0        M3A        North York                                     Parkwoods
1        M4A        North York                              Victoria Village
2        M5A  Downtown Toronto                    Regent Park / Harbourfront
3        M6A        North York             Lawrence Manor / Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

##### Notice the rows have been filtered as desired from 180 rows to 103 rows.

### Find Duplicates

In [9]:
dup1 = df[df.duplicated(['PostalCode'])]
dup1.shape
dup1.head()

Empty DataFrame
Columns: [PostalCode, Borough, Neighborhood]
Index: []

#### These are no duplicate rows with same postal code.

#### As one postal code may contain more than one Neighborhood sepearted by "/".We need to repnace "/" with comma in Neighborhood column.

In [10]:
df['Neighborhood'] = df['Neighborhood'].str.replace('/',',')


#### Check if any neighbourhood which is not assigned.

In [11]:
df_x = df[df.Neighborhood == "Not assigned"]
df_x.head()

Empty DataFrame
Columns: [PostalCode, Borough, Neighborhood]
Index: []

##### There is no such neighbourhood.

#### The Final output will like as follows as expected.

In [12]:
df.head(15)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   
12        M1C       Scarborough   
13        M3C        North York   
14        M4C         East York   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                          Regent Park , Harbourfront  
3                   Lawrence Manor , Lawrence Heights  
4        Queen's Park , Ontario Provincial Government  
5                                    Islington Avenue  
6                                     Malvern , Rouge  
7                                           Don Mills  
8                    Parkview Hill , Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park , Princess Gardens , Martin Gr...  
12           Rouge Hill , Port Union , Highland Creek  
13                                          Don Mills  
14                                   Woodbine Heights

In [13]:
df.shape
x = df.shape[0]
x

103

In [14]:
print("Number of rows in dataframe : ", df.shape[0])

Number of rows in dataframe :  103


### We got our first dataframe as df which has PostalCode, borough and neighbourhood.

### Now, let's get our second dataframe.


#### We will get this through a csv file from the link.

In [15]:
df1 = pd.read_csv("http://cocl.us/Geospatial_data")
df1.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

### Check how many rows both dataframes have.

In [16]:
df.shape
df1.shape
print("Number of rows of df: " , df.shape[0])
print("Number of rows of df1: " , df1.shape[0])

Number of rows of df:  103
Number of rows of df1:  103


#### Notice both have equal rows.


### Now, rename the column of df1 to PostalCode.

In [20]:
df1.rename(columns = {'Postal Code':'PostalCode'}, inplace = True) 
df1.head()


PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

In [21]:
df.head()

PostalCode           Borough                                  Neighborhood
0        M3A        North York                                     Parkwoods
1        M4A        North York                              Victoria Village
2        M5A  Downtown Toronto                    Regent Park , Harbourfront
3        M6A        North York             Lawrence Manor , Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park , Ontario Provincial Government

### Now, merge both dataframes df and df1 on column PostalCode.

In [22]:
df3 = pd.merge(df, df1, on='PostalCode')
df3.head()


PostalCode           Borough                                  Neighborhood  \
0        M3A        North York                                     Parkwoods   
1        M4A        North York                              Victoria Village   
2        M5A  Downtown Toronto                    Regent Park , Harbourfront   
3        M6A        North York             Lawrence Manor , Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park , Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

### Check how many rows in merged dataframe d3

In [23]:
df3.shape
print("Number of rows in merged dataframe d3 : " , df3.shape[0])

Number of rows in merged dataframe d3 :  103


### This is same as df and df1.

### Let's go to third question now. Here we need to explore and cluster the neighborhoods in Toronto. Pick only boroughs with Toronto. Generate map to visualize clusters.

#### Get all dependencies we need.

In [24]:
!conda install -c conda-forge geopy --yes # uncomment this line if not installed before.
!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if not installed before.

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2020.4.5.1 |       hecc5488_0         146 KB  conda-forge
    geopy-1.21.0               |             py_0          58 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    openssl-1.1.1f             |       h516909a_0         2.1 MB  conda-forge
    certifi-2020.4.5.1         |   py36h9f0ad1d_0         151 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0         conda-forge
    geopy:           1

In [27]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files


from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')


Libraries imported.


# let's explore Scarborough.

### Get coordinates of Scarborough  first.

In [62]:

address = 'Scarborough,Toronto'
geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Scarborough are {}, {}.'.format(latitude, longitude))




The geograpical coordinate of Scarborough are 43.773077, -79.257774.


### Create map of Scarborough first

In [63]:
# create map of Scarborough using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)


### Slice the dataframe to get only Scarborough in column Borough.

In [64]:
neighborhoods = df3[df3['Borough'] == 'Scarborough'].reset_index(drop=True)
neighborhoods.head()


PostalCode      Borough                              Neighborhood  \
0        M1B  Scarborough                           Malvern , Rouge   
1        M1C  Scarborough  Rouge Hill , Port Union , Highland Creek   
2        M1E  Scarborough       Guildwood , Morningside , West Hill   
3        M1G  Scarborough                                    Woburn   
4        M1H  Scarborough                                 Cedarbrae   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476

### Check how many rows with Scarborough

In [65]:
neighborhoods.shape
print("Number of rows with Scarborough : ", neighborhoods.shape[0] )


Number of rows with Scarborough :  17


### Superimpose neighborhoods on top of Scarborough map by adding markers.

In [66]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto


### Use Foursquare API to explore the neighborhoods and segment them.


### Define Foursquare Credentials and Version


In [67]:
CLIENT_ID = 'AED1YJAEJO1BSJVJJ5KT2Y2BDGDVPCB21WUXERHTA4B4XXKC' # My Foursquare ID
CLIENT_SECRET = '5WF0PKUAGWG30ASKZSBV12XHC2IDIQLEHRRPU541XI2JVQCH' # My Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('My credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)


My credentails:
CLIENT_ID: AED1YJAEJO1BSJVJJ5KT2Y2BDGDVPCB21WUXERHTA4B4XXKC
CLIENT_SECRET:5WF0PKUAGWG30ASKZSBV12XHC2IDIQLEHRRPU541XI2JVQCH


### Again take a look at Scarborogh datafarme again

In [34]:
neighborhoods.head()

PostalCode      Borough                              Neighborhood  \
0        M1B  Scarborough                           Malvern , Rouge   
1        M1C  Scarborough  Rouge Hill , Port Union , Highland Creek   
2        M1E  Scarborough       Guildwood , Morningside , West Hill   
3        M1G  Scarborough                                    Woburn   
4        M1H  Scarborough                                 Cedarbrae   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476

### Let's get the fourth neighborhood name

In [68]:
neighborhoods.loc[3, 'Neighborhood']



'Woburn'

### It is Woburn. Get Woburn's latitude and longitude.

In [69]:
neighborhood_latitude = neighborhoods.loc[3, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = neighborhoods.loc[3, 'Longitude'] # neighborhood longitude value

neighborhood_name = neighborhoods.loc[3, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))




Latitude and longitude values of Woburn are 43.7709921, -79.21691740000001.


### Get the top 100 venues that are in Woburn within a radius of 500 meters.



### create the GET request URL. Name URL name as url.



In [70]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url 


'https://api.foursquare.com/v2/venues/explore?&client_id=AED1YJAEJO1BSJVJJ5KT2Y2BDGDVPCB21WUXERHTA4B4XXKC&client_secret=5WF0PKUAGWG30ASKZSBV12XHC2IDIQLEHRRPU541XI2JVQCH&v=20180605&ll=43.7709921,-79.21691740000001&radius=500&limit=100'

#### Check results of  GET request.

In [71]:

results = requests.get(url).json()
results


{'meta': {'code': 200, 'requestId': '5e9201dc47e0d6001b56e340'},
 'response': {'headerLocation': 'Toronto',
  'headerFullLocation': 'Toronto',
  'headerLocationGranularity': 'city',
  'totalResults': 4,
  'suggestedBounds': {'ne': {'lat': 43.7754921045, 'lng': -79.21069729639068},
   'sw': {'lat': 43.7664920955, 'lng': -79.22313750360935}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4cc1d28c06c254815ac18547',
       'name': 'Starbucks',
       'location': {'address': '300 Borough Dr',
        'crossStreet': 'Scarborough Town Centre',
        'lat': 43.770037201625215,
        'lng': -79.22115586641958,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.770037201625215,
          'lng': -79.22115586641958}],
        'distance': 356,
        'cc': 'CA

### Define a function that extracts the category of the venue.


In [72]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']


### Clean the json and convert it into a pandas dataframe.


In [73]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()


name         categories        lat        lng
0                Starbucks        Coffee Shop  43.770037 -79.221156
1              Tim Hortons        Coffee Shop  43.770827 -79.223078
2       Korean Grill House  Korean Restaurant  43.770812 -79.214502
3  Al-Hamd Biryani & Pizza  Indian Restaurant  43.767585 -79.219570

In [74]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))


4 venues were returned by Foursquare.


## Explore Neighborhoods in Scarborough in similar fashion. 

### By creating a function to repeat the same process to all the neighborhoods in Scarborough


In [75]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)



## Run the above function on each neighborhood and create a new dataframe called woburn_venues.

In [76]:

woburn_venues = getNearbyVenues(names=neighborhoods['Neighborhood'],
                                   latitudes=neighborhoods['Latitude'],
                                   longitudes=neighborhoods['Longitude']
                                  )


Malvern , Rouge
Rouge Hill , Port Union , Highland Creek
Guildwood , Morningside , West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park , Ionview , East Birchmount Park
Golden Mile , Clairlea , Oakridge
Cliffside , Cliffcrest , Scarborough Village West
Birch Cliff , Cliffside West
Dorset Park , Wexford Heights , Scarborough Town Centre
Wexford , Maryvale
Agincourt
Clarks Corners , Tam O'Shanter , Sullivan
Milliken , Agincourt North , Steeles East , L'Amoreaux East
Steeles West , L'Amoreaux West
Upper Rouge


### Let's check the size 


In [77]:
print(woburn_venues.shape)
woburn_venues.head()


(95, 7)


Neighborhood  Neighborhood Latitude  \
0                           Malvern , Rouge              43.806686   
1                           Malvern , Rouge              43.806686   
2  Rouge Hill , Port Union , Highland Creek              43.784535   
3  Rouge Hill , Port Union , Highland Creek              43.784535   
4  Rouge Hill , Port Union , Highland Creek              43.784535   

   Neighborhood Longitude                         Venue  Venue Latitude  \
0              -79.194353                       Wendy’s       43.807448   
1              -79.194353  T Hamilton & Son Roofing Inc       43.807985   
2              -79.160497             RIGHT WAY TO GOLF       43.785177   
3              -79.160497   Great Shine Window Cleaning       43.783145   
4              -79.160497         Royal Canadian Legion       43.782533   

   Venue Longitude              Venue Category  
0       -79.199056        Fast Food Restaurant  
1       -79.198194  Construction & Landscaping  
2       -79.161108                 Golf Course  
3       -79.157431                Home Service  
4       -79.163085                         Bar

In [78]:
woburn_venues.groupby('Neighborhood').count()


Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               5   
Birch Cliff , Cliffside West                                            4   
Cedarbrae                                                               9   
Clarks Corners , Tam O'Shanter , Sullivan                              14   
Cliffside , Cliffcrest , Scarborough Village West                       3   
Dorset Park , Wexford Heights , Scarborough Tow...                      6   
Golden Mile , Clairlea , Oakridge                                       9   
Guildwood , Morningside , West Hill                                     7   
Kennedy Park , Ionview , East Birchmount Park                           4   
Malvern , Rouge                                                         2   
Milliken , Agincourt North , Steeles East , L'A...                      2   
Rouge Hill , Port Union , Highland Creek                                4   
Scarborough Village                                                     1   
Steeles West , L'Amoreaux West                                         14   
Wexford , Maryvale                                                      7   
Woburn                                                                  4   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Agincourt                                                                5   
Birch Cliff , Cliffside West                                             4   
Cedarbrae                                                                9   
Clarks Corners , Tam O'Shanter , Sullivan                               14   
Cliffside , Cliffcrest , Scarborough Village West                        3   
Dorset Park , Wexford Heights , Scarborough Tow...                       6   
Golden Mile , Clairlea , Oakridge                                        9   
Guildwood , Morningside , West Hill                                      7   
Kennedy Park , Ionview , East Birchmount Park                            4   
Malvern , Rouge                                                          2   
Milliken , Agincourt North , Steeles East , L'A...                       2   
Rouge Hill , Port Union , Highland Creek                                 4   
Scarborough Village                                                      1   
Steeles West , L'Amoreaux West                                          14   
Wexford , Maryvale                                                       7   
Woburn                                                                   4   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Agincourt                                               5               5   
Birch Cliff , Cliffside West                            4               4   
Cedarbrae                                               9               9   
Clarks Corners , Tam O'Shanter , Sullivan              14              14   
Cliffside , Cliffcrest , Scarborough Village West       3               3   
Dorset Park , Wexford Heights , Scarborough Tow...      6               6   
Golden Mile , Clairlea , Oakridge                       9               9   
Guildwood , Morningside , West Hill                     7               7   
Kennedy Park , Ionview , East Birchmount Park           4               4   
Malvern , Rouge                                         2               2   
Milliken , Agincourt North , Steeles East , L'A...      2               2   
Rouge Hill , Port Union , Highland Creek                4               4   
Scarborough Village                                     1               1   
Steeles West , L'Amoreaux West                         14              14   
Wexford , Maryvale          

### Let's find out how many unique categories found in Scarborough.

In [79]:
print('There are {} uniques categories.'.format(len(woburn_venues['Venue Category'].unique())))


There are 57 uniques categories.


### Analyze Each Neighborhood of Scarborough.


In [83]:
# one hot encoding
woburn_onehot = pd.get_dummies(woburn_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
woburn_onehot['Neighborhood'] = woburn_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [woburn_onehot.columns[-1]] + list(woburn_onehot.columns[:-1])
woburn_onehot = woburn_onehot[fixed_columns]


woburn_onehot.shape
woburn_onehot.dtypes
woburn_onehot.head()

Neighborhood  American Restaurant  \
0                           Malvern , Rouge                    0   
1                           Malvern , Rouge                    0   
2  Rouge Hill , Port Union , Highland Creek                    0   
3  Rouge Hill , Port Union , Highland Creek                    0   
4  Rouge Hill , Port Union , Highland Creek                    0   

   Athletics & Sports  Auto Garage  Bakery  Bank  Bar  Breakfast Spot  \
0                   0            0       0     0    0               0   
1                   0            0       0     0    0               0   
2                   0            0       0     0    0               0   
3                   0            0       0     0    0               0   
4                   0            0       0     0    1               0   

   Bus Line  Café  Camera Store  Caribbean Restaurant  Chinese Restaurant  \
0         0     0             0                     0                   0   
1         0     0             0                     0                   0   
2         0     0             0                     0                   0   
3         0     0             0                     0                   0   
4         0     0             0                     0                   0   

   Clothing Store  Coffee Shop  College Stadium  Construction & Landscaping  \
0               0            0                0                           0   
1               0            0                0                           1   
2               0            0                0                           0   
3               0            0                0                           0   
4               0            0                0                           0   

   Convenience Store  Department Store  Electronics Store  \
0                  0                 0                  0   
1                  0                 0                  0   
2                  0                 0                  0   
3                  0                 0                  0   
4                  0                 0                  0   

   Fast Food Restaurant  Fried Chicken Joint  Gas Station  \
0                     1                    0            0   
1                     0                    0            0   
2                     0                    0            0   
3                     0                    0            0   
4                     0                    0            0   

   General Entertainment  Golf Course  Grocery Store  Hakka Restaurant  \
0                      0            0              0                 0   
1                      0            0              0                 0   
2                      0            1              0                 0   
3                      0            0              0                 0   
4                      0            0              0                 0   

   History Museum  Home Service  Ice Cream Shop  Indian Restaurant  \
0               0             0               0                  0   
1               0             0               0                  0   
2               0             0               0                  0   
3               0             1               0                  0   
4               0             0               0                  0   

   Intersection  Italian Restaurant  Korean Restaurant  \
0             0                   0                  0   
1             0                   0                  0   
2             0                   0                  0   
3             0                   0                  0   
4             0                   0                  0   

   Latin American Restaurant  Light Rail Station  Lounge  Medical Center  \
0                          0                   0       0               0   
1                          0                   0       0               0   
2                          0                   0       0            

### check size.


In [48]:
woburn_onehot.shape

(95, 58)

In [84]:
woburn_onehot.head()

Neighborhood  American Restaurant  \
0                           Malvern , Rouge                    0   
1                           Malvern , Rouge                    0   
2  Rouge Hill , Port Union , Highland Creek                    0   
3  Rouge Hill , Port Union , Highland Creek                    0   
4  Rouge Hill , Port Union , Highland Creek                    0   

   Athletics & Sports  Auto Garage  Bakery  Bank  Bar  Breakfast Spot  \
0                   0            0       0     0    0               0   
1                   0            0       0     0    0               0   
2                   0            0       0     0    0               0   
3                   0            0       0     0    0               0   
4                   0            0       0     0    1               0   

   Bus Line  Café  Camera Store  Caribbean Restaurant  Chinese Restaurant  \
0         0     0             0                     0                   0   
1         0     0             0                     0                   0   
2         0     0             0                     0                   0   
3         0     0             0                     0                   0   
4         0     0             0                     0                   0   

   Clothing Store  Coffee Shop  College Stadium  Construction & Landscaping  \
0               0            0                0                           0   
1               0            0                0                           1   
2               0            0                0                           0   
3               0            0                0                           0   
4               0            0                0                           0   

   Convenience Store  Department Store  Electronics Store  \
0                  0                 0                  0   
1                  0                 0                  0   
2                  0                 0                  0   
3                  0                 0                  0   
4                  0                 0                  0   

   Fast Food Restaurant  Fried Chicken Joint  Gas Station  \
0                     1                    0            0   
1                     0                    0            0   
2                     0                    0            0   
3                     0                    0            0   
4                     0                    0            0   

   General Entertainment  Golf Course  Grocery Store  Hakka Restaurant  \
0                      0            0              0                 0   
1                      0            0              0                 0   
2                      0            1              0                 0   
3                      0            0              0                 0   
4                      0            0              0                 0   

   History Museum  Home Service  Ice Cream Shop  Indian Restaurant  \
0               0             0               0                  0   
1               0             0               0                  0   
2               0             0               0                  0   
3               0             1               0                  0   
4               0             0               0                  0   

   Intersection  Italian Restaurant  Korean Restaurant  \
0             0                   0                  0   
1             0                   0                  0   
2             0                   0                  0   
3             0                   0                  0   
4             0                   0                  0   

   Latin American Restaurant  Light Rail Station  Lounge  Medical Center  \
0                          0                   0       0               0   
1                          0                   0       0               0   
2                          0                   0       0            

## Group rows by neighborhood and by taking the mean of the frequency of occurrence of each category


In [85]:
woburn_grouped = woburn_onehot.groupby('Neighborhood').mean().reset_index()
woburn_grouped


Neighborhood  American Restaurant  \
0                                           Agincourt             0.000000   
1                        Birch Cliff , Cliffside West             0.000000   
2                                           Cedarbrae             0.000000   
3           Clarks Corners , Tam O'Shanter , Sullivan             0.000000   
4   Cliffside , Cliffcrest , Scarborough Village West             0.333333   
5   Dorset Park , Wexford Heights , Scarborough To...             0.000000   
6                   Golden Mile , Clairlea , Oakridge             0.000000   
7                 Guildwood , Morningside , West Hill             0.000000   
8       Kennedy Park , Ionview , East Birchmount Park             0.000000   
9                                     Malvern , Rouge             0.000000   
10  Milliken , Agincourt North , Steeles East , L'...             0.000000   
11           Rouge Hill , Port Union , Highland Creek             0.000000   
12                                Scarborough Village             0.000000   
13                     Steeles West , L'Amoreaux West             0.000000   
14                                 Wexford , Maryvale             0.000000   
15                                             Woburn             0.000000   

    Athletics & Sports  Auto Garage    Bakery      Bank   Bar  Breakfast Spot  \
0             0.000000     0.000000  0.000000  0.000000  0.00        0.200000   
1             0.000000     0.000000  0.000000  0.000000  0.00        0.000000   
2             0.111111     0.000000  0.111111  0.111111  0.00        0.000000   
3             0.000000     0.000000  0.000000  0.071429  0.00        0.000000   
4             0.000000     0.000000  0.000000  0.000000  0.00        0.000000   
5             0.000000     0.000000  0.000000  0.000000  0.00        0.000000   
6             0.000000     0.000000  0.222222  0.000000  0.00        0.000000   
7             0.000000     0.000000  0.000000  0.142857  0.00        0.142857   
8             0.000000     0.000000  0.000000  0.000000  0.00        0.000000   
9             0.000000     0.000000  0.000000  0.000000  0.00        0.000000   
10            0.000000     0.000000  0.000000  0.000000  0.00        0.000000   
11            0.000000     0.000000  0.000000  0.000000  0.25        0.000000   
12            0.000000     0.000000  0.000000  0.000000  0.00        0.000000   
13            0.000000     0.000000  0.000000  0.071429  0.00        0.071429   
14            0.000000     0.142857  0.142857  0.000000  0.00        0.142857   
15            0.000000     0.000000  0.000000  0.000000  0.00        0.000000   

    Bus Line  Café  Camera Store  Caribbean Restaurant  Chinese Restaurant  \
0   0.000000  0.00      0.000000              0.000000            0.000000   
1   0.000000  0.25      0.000000              0.000000            0.000000   
2   0.000000  0.00      0.000000              0.111111            0.000000   
3   0.000000  0.00      0.000000              0.000000            0.071429   
4   0.000000  0.00      0.000000              0.000000            0.000000   
5   0.000000  0.00      0.000000              0.000000            0.166667   
6   0.222222  0.00      0.000000              0.000000            0.000000   
7   0.000000  0.00      0.000000              0.000000            0.000000   
8   0.000000  0.00      0.000000              0.000000            0.250000   
9   0.000000  0.00      0.000000              0.000000            0.000000   
10  0.000000  0.00      0.000000              0.000000            0.000000   
11  0.000000  0.00      0.000000              0.000000            0.000000   
12  0.000000  0.00      0.000000              0.000000            0.000000   
13  0.000000  0.00      0.071429              0.000000            0.142857   
14  0.000000  0.00      0.000000              0.000000            0.000000   
15  0.000000  0.00      0.000000              0.000000            0.000000   

    Cloth

In [86]:
woburn_grouped.shape

(16, 58)

### Let's print each neighborhood along with the top 5 most common venues


In [87]:
num_top_venues = 5

for hood in woburn_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = woburn_grouped[woburn_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')
    
    #Cluster Labels

----Agincourt----
                       venue  freq
0  Latin American Restaurant   0.2
1               Skating Rink   0.2
2             Breakfast Spot   0.2
3                     Lounge   0.2
4             Clothing Store   0.2


----Birch Cliff , Cliffside West----
                   venue  freq
0           Skating Rink  0.25
1                   Café  0.25
2  General Entertainment  0.25
3        College Stadium  0.25
4    American Restaurant  0.00


----Cedarbrae----
                 venue  freq
0      Thai Restaurant  0.11
1               Bakery  0.11
2                 Bank  0.11
3  Fried Chicken Joint  0.11
4               Lounge  0.11


----Clarks Corners , Tam O'Shanter , Sullivan----
                venue  freq
0            Pharmacy  0.14
1         Pizza Place  0.14
2   Convenience Store  0.07
3  Chinese Restaurant  0.07
4         Gas Station  0.07


----Cliffside , Cliffcrest , Scarborough Village West----
                 venue  freq
0  American Restaurant  0.33
1              

### Putting that into a pandas dataframe.Writing a function to sort the venues in descending order.


In [88]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

#Cluster Labels

### Creating the new dataframe and display the top 10 venues for each neighborhood.


In [89]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = woburn_grouped['Neighborhood']

for ind in np.arange(woburn_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(woburn_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

#Cluster Labels

Neighborhood 1st Most Common Venue  \
0                                          Agincourt          Skating Rink   
1                       Birch Cliff , Cliffside West       College Stadium   
2                                          Cedarbrae       Thai Restaurant   
3          Clarks Corners , Tam O'Shanter , Sullivan           Pizza Place   
4  Cliffside , Cliffcrest , Scarborough Village West   American Restaurant   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot  Latin American Restaurant                Lounge   
1          Skating Rink      General Entertainment                  Café   
2    Athletics & Sports           Hakka Restaurant                Bakery   
3              Pharmacy               Noodle House    Chinese Restaurant   
4         Movie Theater                      Motel         Grocery Store   

   5th Most Common Venue  6th Most Common Venue       7th Most Common Venue  \
0         Clothing Store  Vietnamese Restaurant  Construction & Landscaping   
1  Vietnamese Restaurant       Hakka Restaurant               Grocery Store   
2                   Bank            Gas Station         Fried Chicken Joint   
3        Thai Restaurant            Gas Station               Shopping Mall   
4            Golf Course  General Entertainment                 Gas Station   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0         Grocery Store           Golf Course  General Entertainment  
1           Golf Course           Gas Station    Fried Chicken Joint  
2  Caribbean Restaurant                Lounge      Convenience Store  
3                  Bank   Fried Chicken Joint     Italian Restaurant  
4   Fried Chicken Joint  Fast Food Restaurant      Electronics Store

## Cluster Neighborhoods of Scarborough using k-means.

### Run k-means to cluster the neighborhood into 5 clusters.


In [90]:
# set number of clusters
kclusters = 5

woburn_grouped_clustering = woburn_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(woburn_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 


array([0, 0, 4, 4, 4, 4, 4, 4, 1, 2], dtype=int32)

In [59]:
neighborhoods.head()

PostalCode      Borough                              Neighborhood  \
0        M1B  Scarborough                           Malvern , Rouge   
1        M1C  Scarborough  Rouge Hill , Port Union , Highland Creek   
2        M1E  Scarborough       Guildwood , Morningside , West Hill   
3        M1G  Scarborough                                    Woburn   
4        M1H  Scarborough                                 Cedarbrae   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476

In [91]:
woburn_grouped_clustering.head()

American Restaurant  Athletics & Sports  Auto Garage    Bakery      Bank  \
0             0.000000            0.000000          0.0  0.000000  0.000000   
1             0.000000            0.000000          0.0  0.000000  0.000000   
2             0.000000            0.111111          0.0  0.111111  0.111111   
3             0.000000            0.000000          0.0  0.000000  0.071429   
4             0.333333            0.000000          0.0  0.000000  0.000000   

   Bar  Breakfast Spot  Bus Line  Café  Camera Store  Caribbean Restaurant  \
0  0.0             0.2       0.0  0.00           0.0              0.000000   
1  0.0             0.0       0.0  0.25           0.0              0.000000   
2  0.0             0.0       0.0  0.00           0.0              0.111111   
3  0.0             0.0       0.0  0.00           0.0              0.000000   
4  0.0             0.0       0.0  0.00           0.0              0.000000   

   Chinese Restaurant  Clothing Store  Coffee Shop  College Stadium  \
0            0.000000             0.2          0.0             0.00   
1            0.000000             0.0          0.0             0.25   
2            0.000000             0.0          0.0             0.00   
3            0.071429             0.0          0.0             0.00   
4            0.000000             0.0          0.0             0.00   

   Construction & Landscaping  Convenience Store  Department Store  \
0                         0.0           0.000000               0.0   
1                         0.0           0.000000               0.0   
2                         0.0           0.000000               0.0   
3                         0.0           0.071429               0.0   
4                         0.0           0.000000               0.0   

   Electronics Store  Fast Food Restaurant  Fried Chicken Joint  Gas Station  \
0                0.0              0.000000             0.000000     0.000000   
1                0.0              0.000000             0.000000     0.000000   
2                0.0              0.000000             0.111111     0.111111   
3                0.0              0.071429             0.071429     0.071429   
4                0.0              0.000000             0.000000     0.000000   

   General Entertainment  Golf Course  Grocery Store  Hakka Restaurant  \
0                   0.00          0.0            0.0          0.000000   
1                   0.25          0.0            0.0          0.000000   
2                   0.00          0.0            0.0          0.111111   
3                   0.00          0.0            0.0          0.000000   
4                   0.00          0.0            0.0          0.000000   

   History Museum  Home Service  Ice Cream Shop  Indian Restaurant  \
0             0.0           0.0             0.0                0.0   
1             0.0           0.0             0.0                0.0   
2             0.0           0.0             0.0                0.0   
3             0.0           0.0             0.0                0.0   
4             0.0           0.0             0.0                0.0   

   Intersection  Italian Restaurant  Korean Restaurant  \
0           0.0            0.000000                0.0   
1           0.0            0.000000                0.0   
2           0.0            0.000000                0.0   
3           0.0            0.071429                0.0   
4           0.0            0.000000                0.0   

   Latin American Restaurant  Light Rail Station    Lounge  Medical Center  \
0                        0.2                 0.0  0.200000             0.0   
1                        0.0                 0.0  0.000000             0.0   
2                        0.0                 0.0  0.111111             0.0   
3                        0.0                 0.0  0.000000             0.0   
4                        0.0                 0.0  0.000000             0.0   

   Metro Station  Mexican Restaurant  Middle E

In [61]:
neighborhoods_venues_sorted.head()
neighborhoods_venues_sorted.dtypes

Neighborhood              object
1st Most Common Venue     object
2nd Most Common Venue     object
3rd Most Common Venue     object
4th Most Common Venue     object
5th Most Common Venue     object
6th Most Common Venue     object
7th Most Common Venue     object
8th Most Common Venue     object
9th Most Common Venue     object
10th Most Common Venue    object
dtype: object

### Create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood. For this mege two dataframes. neighborhoods & neighborhoods_venues_sorted. The merger will be done on common column Neighborhood.


In [92]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

woburn_merged = neighborhoods

# merge woburn_merged with neighborhoods_venues_sorted to add latitude/longitude for each neighborhood
woburn_merged = woburn_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

woburn_merged.head() # check the last columns!


PostalCode      Borough                              Neighborhood  \
0        M1B  Scarborough                           Malvern , Rouge   
1        M1C  Scarborough  Rouge Hill , Port Union , Highland Creek   
2        M1E  Scarborough       Guildwood , Morningside , West Hill   
3        M1G  Scarborough                                    Woburn   
4        M1H  Scarborough                                 Cedarbrae   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.806686 -79.194353             2.0  Fast Food Restaurant   
1  43.784535 -79.160497             4.0        History Museum   
2  43.763573 -79.188711             4.0          Intersection   
3  43.770992 -79.216917             1.0           Coffee Shop   
4  43.773136 -79.239476             4.0       Thai Restaurant   

        2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0  Construction & Landscaping  Vietnamese Restaurant          Home Service   
1                 Golf Course                    Bar          Home Service   
2                        Bank    Rental Car Location        Breakfast Spot   
3           Indian Restaurant      Korean Restaurant       College Stadium   
4          Athletics & Sports       Hakka Restaurant                Bakery   

  5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
0      Hakka Restaurant         Grocery Store            Golf Course   
1       College Stadium      Hakka Restaurant          Grocery Store   
2     Electronics Store        Medical Center     Mexican Restaurant   
3         Grocery Store           Golf Course  General Entertainment   
4                  Bank           Gas Station    Fried Chicken Joint   

   8th Most Common Venue       9th Most Common Venue 10th Most Common Venue  
0  General Entertainment                 Gas Station    Fried Chicken Joint  
1  General Entertainment                 Gas Station    Fried Chicken Joint  
2  Vietnamese Restaurant  Construction & Landscaping            Golf Course  
3            Gas Station         Fried Chicken Joint   Fast Food Restaurant  
4   Caribbean Restaurant                      Lounge      Convenience Store

### Check datatype of Cluster Labels



In [93]:
woburn_merged.dtypes

PostalCode                 object
Borough                    object
Neighborhood               object
Latitude                  float64
Longitude                 float64
Cluster Labels            float64
1st Most Common Venue      object
2nd Most Common Venue      object
3rd Most Common Venue      object
4th Most Common Venue      object
5th Most Common Venue      object
6th Most Common Venue      object
7th Most Common Venue      object
8th Most Common Venue      object
9th Most Common Venue      object
10th Most Common Venue     object
dtype: object

In [94]:
woburn_merged['Cluster Labels']

0     2.0
1     4.0
2     4.0
3     1.0
4     4.0
5     3.0
6     1.0
7     4.0
8     4.0
9     0.0
10    4.0
11    4.0
12    0.0
13    4.0
14    3.0
15    4.0
16    NaN
Name: Cluster Labels, dtype: float64

In [95]:
woburn_merged.shape #(17, 16)

(17, 16)

### Slice to remove last row with NaN value

In [96]:
print("woburn_merged:bef:" , woburn_merged.shape)
woburn_merged = woburn_merged.iloc[0:16, 0:17]
print("woburn_merged:aft:" , woburn_merged.shape)
woburn_merged['Cluster Labels']

woburn_merged:bef: (17, 16)
woburn_merged:aft: (16, 16)


0     2.0
1     4.0
2     4.0
3     1.0
4     4.0
5     3.0
6     1.0
7     4.0
8     4.0
9     0.0
10    4.0
11    4.0
12    0.0
13    4.0
14    3.0
15    4.0
Name: Cluster Labels, dtype: float64

### Convert to integer and check datatype of this field.

In [97]:
woburn_merged['Cluster Labels'] = woburn_merged['Cluster Labels'].apply(np.int)
woburn_merged.dtypes
#woburn_merged['Cluster Labels']

PostalCode                 object
Borough                    object
Neighborhood               object
Latitude                  float64
Longitude                 float64
Cluster Labels              int64
1st Most Common Venue      object
2nd Most Common Venue      object
3rd Most Common Venue      object
4th Most Common Venue      object
5th Most Common Venue      object
6th Most Common Venue      object
7th Most Common Venue      object
8th Most Common Venue      object
9th Most Common Venue      object
10th Most Common Venue     object
dtype: object

In [98]:
woburn_merged['Cluster Labels']

0     2
1     4
2     4
3     1
4     4
5     3
6     1
7     4
8     4
9     0
10    4
11    4
12    0
13    4
14    3
15    4
Name: Cluster Labels, dtype: int64

### Visualize the resulting clusters.


In [99]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(woburn_merged['Latitude'], woburn_merged['Longitude'], woburn_merged['Neighborhood'], woburn_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters


## Examine Clusters


### Examine each cluster and determine the discriminating venue categories that distinguish each cluster. 

# Cluster 1


In [100]:
woburn_merged.loc[woburn_merged['Cluster Labels'] == 0, woburn_merged.columns[[1] + list(range(5, woburn_merged.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
9   Scarborough               0       College Stadium          Skating Rink   
12  Scarborough               0          Skating Rink        Breakfast Spot   

        3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
9       General Entertainment                  Café  Vietnamese Restaurant   
12  Latin American Restaurant                Lounge         Clothing Store   

    6th Most Common Venue       7th Most Common Venue 8th Most Common Venue  \
9        Hakka Restaurant               Grocery Store           Golf Course   
12  Vietnamese Restaurant  Construction & Landscaping         Grocery Store   

   9th Most Common Venue 10th Most Common Venue  
9            Gas Station    Fried Chicken Joint  
12           Golf Course  General Entertainment

# Cluster 2

In [101]:
woburn_merged.loc[woburn_merged['Cluster Labels'] == 1, woburn_merged.columns[[1] + list(range(5, woburn_merged.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
3  Scarborough               1           Coffee Shop     Indian Restaurant   
6  Scarborough               1            Playground      Department Store   

  3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
3     Korean Restaurant       College Stadium          Grocery Store   
6    Chinese Restaurant           Coffee Shop  Vietnamese Restaurant   

  6th Most Common Venue  7th Most Common Venue 8th Most Common Venue  \
3           Golf Course  General Entertainment           Gas Station   
6       College Stadium       Hakka Restaurant         Grocery Store   

  9th Most Common Venue 10th Most Common Venue  
3   Fried Chicken Joint   Fast Food Restaurant  
6           Golf Course  General Entertainment

# Cluster 3


In [102]:
woburn_merged.loc[woburn_merged['Cluster Labels'] == 2, woburn_merged.columns[[1] + list(range(5, woburn_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0  Scarborough               2  Fast Food Restaurant   

        2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0  Construction & Landscaping  Vietnamese Restaurant          Home Service   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0      Hakka Restaurant         Grocery Store           Golf Course   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0  General Entertainment           Gas Station    Fried Chicken Joint

# Cluster 4

In [103]:
woburn_merged.loc[woburn_merged['Cluster Labels'] == 3, woburn_merged.columns[[1] + list(range(5, woburn_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  2nd Most Common Venue  \
5   Scarborough               3            Playground  Vietnamese Restaurant   
14  Scarborough               3            Playground                   Park   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
5             Coffee Shop      Hakka Restaurant         Grocery Store   
14  Vietnamese Restaurant           Coffee Shop         Grocery Store   

   6th Most Common Venue  7th Most Common Venue 8th Most Common Venue  \
5            Golf Course  General Entertainment           Gas Station   
14           Golf Course  General Entertainment           Gas Station   

   9th Most Common Venue 10th Most Common Venue  
5    Fried Chicken Joint   Fast Food Restaurant  
14   Fried Chicken Joint   Fast Food Restaurant

# Cluster 5

In [104]:
woburn_merged.loc[woburn_merged['Cluster Labels'] == 4, woburn_merged.columns[[1] + list(range(5, woburn_merged.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
1   Scarborough               4             History Museum   
2   Scarborough               4               Intersection   
4   Scarborough               4            Thai Restaurant   
7   Scarborough               4                     Bakery   
8   Scarborough               4        American Restaurant   
10  Scarborough               4          Indian Restaurant   
11  Scarborough               4  Middle Eastern Restaurant   
13  Scarborough               4                Pizza Place   
15  Scarborough               4       Fast Food Restaurant   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1             Golf Course                   Bar          Home Service   
2                    Bank   Rental Car Location        Breakfast Spot   
4      Athletics & Sports      Hakka Restaurant                Bakery   
7                Bus Line        Ice Cream Shop         Metro Station   
8           Movie Theater                 Motel         Grocery Store   
10  Vietnamese Restaurant    Light Rail Station             Pet Store   
11                 Bakery        Sandwich Place         Shopping Mall   
13               Pharmacy          Noodle House    Chinese Restaurant   
15     Chinese Restaurant          Camera Store           Pizza Place   

   5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
1        College Stadium       Hakka Restaurant         Grocery Store   
2      Electronics Store         Medical Center    Mexican Restaurant   
4                   Bank            Gas Station   Fried Chicken Joint   
7           Soccer Field           Intersection                  Park   
8            Golf Course  General Entertainment           Gas Station   
10    Chinese Restaurant        College Stadium         Grocery Store   
11        Breakfast Spot            Auto Garage         Grocery Store   
13       Thai Restaurant            Gas Station         Shopping Mall   
15           Coffee Shop             Nail Salon        Breakfast Spot   

    8th Most Common Venue       9th Most Common Venue 10th Most Common Venue  
1   General Entertainment                 Gas Station    Fried Chicken Joint  
2   Vietnamese Restaurant  Construction & Landscaping            Golf Course  
4    Caribbean Restaurant                      Lounge      Convenience Store  
7       Electronics Store           Convenience Store       Department Store  
8     Fried Chicken Joint        Fast Food Restaurant      Electronics Store  
10            Golf Course       General Entertainment            Gas Station  
11            Golf Course       General Entertainment        College Stadium  
13                   Bank         Fried Chicken Joint     Italian Restaurant  
15               Pharmacy               Grocery Store         Sandwich Place

### This concludes assignemnet.Thank you.